### PR TITLE
Add html5 validations

### PR DESCRIPTION
**Pull request by dbernheisel**
- By default, add validation options if the form source is a changeset with validations.
- Allow `validate: false` option to opt out of html5 validations.
- Allow `validate_regex: false` option to opt out of html5 pattern validation.
- Rewrite the tests
- http://caniuse.com/#feat=form-validation . the last holdout for form validations was iOS Safari, which now provides visual feedback for invalid forms. These browsers also implement the `:invalid` and `:valid` and `:required` CSS pseudo-selectors to help with styling.

### CODE REVIEW

File: lib/formulator.ex [hunk 1]
```diff
@@ -26,13 +26,23 @@ defmodule Formulator do
       #=> <input id="user_name" name="user[email]" aria-label="email" type="text" value="">
 
   Passing other options:
-      <%= input form, :email, label: [class: "control-label"] %>
+      <%= input form, :email, as: :email, label: [class: "control-label"] %>
```
Comment (drapergeek) [inline lib/formulator.ex]: This is to show the `label` part, not the `as:`. For now I'd say let's either leave it alone or if it really bothers you, change the email to be "name".

File: lib/formulator.ex [hunk 2]
```diff
@@ -26,13 +26,23 @@ defmodule Formulator do
...
-      #=> <input id="user_email" name="user[email]" type="text" value="">
+      #=> <input id="user_email" type="email" name="user[email]" value="">
 
-      <%= input form, :email, class: "my-email-class", label: [class: "my-email-label-class"] %>
+      <%= input form, :email, as: :email, placeholder: "your@email.com", class: "my-email-class", label: [class: "my-email-label-class"] %>
       #=> <label class="my-email-label-class" for="user_email">Email</label>
-      #=> <input id="user_email" name="user[email]" type="text" value="" class="my-email-class">
+      #=> <input id="user_email" type="email" name="user[email]" placeholder: "your@email.com" value="" class="my-email-class">
+
+  With validation attributes:
```
Comment (drapergeek) [inline lib/formulator.ex]: Seems like we need more explanation of what is happening exactly. From the outside it's not clear that this requires a changeset that has an applicable validation.

File: lib/formulator.ex [hunk 3]
```diff
@@ -83,16 +93,37 @@ defmodule Formulator do
   defp build_input(form, field, options, error) do
     input_type = options[:as] || :text
     input_class = options[:class] || ""
+
     options =
       options
+      |> add_validation_attributes(form, field)
```
Comment (drapergeek) [inline lib/formulator.ex]: What do you think about moving the associated deleting of `:validate` into this function? Seems like this function should handle its own cleanup.

File: lib/formulator.ex [hunk 4]
```diff
@@ -83,16 +93,37 @@ defmodule Formulator do
...
       |> Keyword.delete(:as)
+      |> Keyword.delete(:validate)
+      |> Keyword.delete(:validate_regex)
       |> Keyword.put(:class, add_error_class(input_class, error.class))
 
     apply(Phoenix.HTML.Form, input_function(input_type), [form, field, options])
   end
 
+  defp add_validation_attributes([validate: true] = options, form, field) do
+    Phoenix.HTML.Form.input_validations(form, field)
```
Comment (drapergeek) [inline lib/formulator.ex]: What do you think about starting the pipe here with a bare term?

`form |> Phoenix.html |> `

File: test/formulator_test.exs [hunk 5]
```diff
@@ -2,50 +2,159 @@ defmodule FormulatorTest do
...
+        |> extract_field(:first)
 
       assert label |> to_string == ~s(<label for="_name">Customer Name</label>)
     end
 
     test "passing `label: false` will add `aria-label` to the input" do
-      [{:safe, input} | _] = %Form{data: %{last_name: ""}}
+      input =
+        %{last_name: ""}
+        |> prepare_form()
```
Comment (drapergeek) [inline test/formulator_test.exs]: You don't need empty parens if it's being piped.

File: test/formulator_test.exs [hunk 6]
```diff
@@ -2,50 +2,159 @@ defmodule FormulatorTest do
...
+        %{name: "Fluff"}
+        |> prepare_changeset_form(:novalidate)
+        |> Formulator.input(:name, validate: true)
+        |> extract_field(:second)
+
+      refute input |> to_string =~ ~s(required)
+    end
+  end
+
+  describe "input - validation - number" do
```
Comment (drapergeek) [inline test/formulator_test.exs]: Do we need tests here? Isn't this testing what phoenix gives to us?

Comment (drapergeek) [inline test/formulator_test.exs]: (checking the number in addition to the required is what I mean).

Comment (dbernheisel) [inline test/formulator_test.exs]: Yes you're right. It's going through the same code. I'll remove it.

File: test/formulator_test.exs (hunk 5 above)
Comment (dbernheisel) [inline test/formulator_test.exs]: 👍

File: lib/formulator.ex (hunk 4 above)
Comment (dbernheisel) [inline lib/formulator.ex]: 👍

File: lib/formulator.ex (hunk 3 above)
Comment (dbernheisel) [inline lib/formulator.ex]: 👍

File: lib/formulator.ex (hunk 1 above)
Comment (dbernheisel) [inline lib/formulator.ex]: 👍

File: lib/formulator.ex [hunk 7]
```diff
@@ -83,16 +116,60 @@ defmodule Formulator do
...
+      |> Keyword.merge(options)
+      |> Keyword.delete(:validate)
+    else
+      options
+    end
+  end
+  defp add_validation_attributes(options, _, _), do: options
+
+  defp add_format_validation_attribute(options, %{impl: impl, source: %{}} = form, field) when is_atom(impl) do
+    with true <- should_validate_regex?(options),
```
Comment (drapergeek) [inline lib/formulator.ex]: Could this be simplified with an `if` over the `with`?

File: lib/formulator.ex [hunk 8]
```diff
@@ -83,16 +116,60 @@ defmodule Formulator do
...
+      |> Keyword.delete(:validate_regex)
+    else
+      _ -> options
+    end
+  end
+  defp add_format_validation_attribute(options, _, _), do: options
+
+  defp should_validate?(options) do
+    Enum.any?([
+      options[:validate] == true,
```
Comment (drapergeek) [inline lib/formulator.ex]: Do you need the `== true` here for these?

Comment (dbernheisel) [inline lib/formulator.ex]: I want to check for the actual boolean and not that it was truthy.

File: README.md [hunk 9]
```diff
@@ -35,8 +35,12 @@ prefer, you can also specify the input type:
 
 ```elixir
   <%= input form, :description, as: :textarea %>
+  <%= input form, :count, as: :number %>
+  <%= input form, :email_address, as: :email %>
 ```
 
+See `Formulator.input/3` for more examples.
```
Comment (drapergeek) [inline README.md]: How about a link?

File: lib/formulator.ex [hunk 10]
```diff
@@ -83,16 +116,60 @@ defmodule Formulator do
...
+      Application.get_env(:formulator, :validate, true) == true,
+    ])
+  end
+
+  defp should_validate_regex?(options) do
+    Enum.any?([
```
Comment (drapergeek) [inline lib/formulator.ex]: This logic is duplicated with the only difference being the key that is used. What do you think about extracting this into something like `option_enabled?(:validate_regex)`

File: mix.exs [hunk 11]
```diff
@@ -9,7 +9,16 @@ defmodule Formulator.Mixfile do
...
      start_permanent: Mix.env == :prod,
      deps: deps(),
-     docs: [source_ref: "#{@version}", extras: ["README.md"], main: "readme"],
+     docs: [
+       source_ref: "#{@version}",
+       extras: [
+         "README.md",
+       ],
+       deps: [
+         ecto: "https://hexdocs.pm/ecto",
```
Comment (drapergeek) [inline mix.exs]: Is this truly a dependency? You don't have to use that feature and the library only requires it in the tests.

File: lib/formulator.ex (hunk 10 above)
Comment (dbernheisel) [reply lib/formulator.ex]: 👍

File: README.md (hunk 9 above)
Comment (dbernheisel) [reply README.md]: 👍

File: mix.exs (hunk 11 above)
Comment (dbernheisel) [reply mix.exs]: Their docs say that I should add these configs to ex_docs to link to other libraries' documentation. I can't seem to make this auto-link even when this configured, so apparently I'm doing something wrong. I'm just going to remove it.

File: lib/formulator.ex [hunk 12]
```diff
@@ -14,25 +14,58 @@ defmodule Formulator do
     When given `false`, does not create a label tag. Instead, an `aria-label`
     attribute is added to the input to improve accessibility.
 
+    * `:validate` - Defaults to application config. When provided a form created with an
+    ecto changeset that contains validations, then Formulator will add HTML5
```
Comment (jsteiner) [inline lib/formulator.ex]: Nitpick - `Ecto` should be capitalized.

File: lib/formulator.ex [hunk 13]
```diff
@@ -14,25 +14,58 @@ defmodule Formulator do
...
   Passing other options:
-      <%= input form, :email, label: [class: "control-label"] %>
-      #=> <label class="control-label" for="user_email">Email</label>
-      #=> <input id="user_email" name="user[email]" type="text" value="">
+      <%= input form, :name, label: [class: "control-label"] %>
+      #=> <label class="control-label" for="user_name">Name</label>
+      #=> <input id="user_name" type="text" name="user[name]" value="">
+
+  Using different input types:
+      <%= input form, :email_address, as: :email, placeholder: "your@email.com", class: "my-email-class", label: [class: "my-email-label-class"] %>
```
Comment (jsteiner) [inline lib/formulator.ex]: Thoughts on collapsing this onto multiple lines?

File: lib/formulator.ex [hunk 14]
```diff
@@ -14,25 +14,58 @@ defmodule Formulator do
...
+      #=> <label class="my-email-label-class" for="user_email_address">Email Address</label>
+      #=> <input id="user_email_address" type="email" name="user[email_address]" placeholder: "your@email.com" value="" class="my-email-class">
```
Comment (jsteiner) [inline lib/formulator.ex]: Likewise

File: lib/formulator.ex [hunk 15]
```diff
@@ -14,25 +14,58 @@ defmodule Formulator do
...
+
+  Or a number input:
+      <%= input form, :count, as: :number %>
```
Comment (jsteiner) [inline lib/formulator.ex]: Should we automatically infer this with:`Phoenix.HTML.Form.input_type/2` ?

Comment (jsteiner) [reply lib/formulator.ex]: Also, if you agree, I'm totally fine pushing that to another PR

File: lib/formulator.ex (hunk 12 above)
Comment (dbernheisel) [reply lib/formulator.ex]: 👍

File: lib/formulator.ex (hunk 15 above)
Comment (dbernheisel) [reply lib/formulator.ex]: It's not related to this PR, so ya, it should go into a different one. I think inferring it makes sense, but I think that'd be an odd behavior if we only did it for numbers. I assume we would infer from an Ecto.Schema, and if we do it for numbers, we should probably do it for booleans, and maybe on emails, and passwords, etc.

**Pull request merged**